File: Word Pattern 290.java
class Solution {// O(n) time for for loop and O(n + k ) time for split
    public boolean wordPattern(String pattern, String s) {
        
        // String[] p = pattern.split("(?!^)");
        String[] str = s.split(" ");
         //Check string lenght equal or not
       if(pattern.length() != str.length) return false;
        
        //creating Hashmap to store mapping
        HashMap<Character, String> mapp = new HashMap<>();
        
        for(int i=0; i<pattern.length() ; i++){ // time complexity of O(n)
            //creating and checking the mapping of each caharter
            char c1 = pattern.charAt(i);
            String c2 = str[i];
            // System.out.println(c1+"--->"+ c2);
            //checking whether the key is in map or not
            if(mapp.containsKey(c1)){
                // System.out.println(mapp.get(c1));
                if(mapp.get(c1).equals(c2)==false){
                   return false ; 
                    
                }
            }else{
                // if map does not contains key but value is present the mapping incorrect
                  if(mapp.containsValue(c2)) return false;
                //key val no present then add a mapping for it
                // System.out.println("put");
                mapp.put(c1,c2);
            }
        }
        return true;
    }
}   

File: Word_Pattern_290.py
# Time Complexity: O(n) since we are traversing the array once 
# Space Complexity: O(n) since we are creating 2 different dictionaries and 1 list 
# Code:
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        
        pattern_dict = {}
        s_dict = {}
        
        n = len(pattern)
        splitted_string = s.split(' ')
        
        m = len(splitted_string)
        
        if n != m:  # Check if the lengths are not equal
            return False
        
        for i in range(n):
            if pattern[i] not in pattern_dict:                      #  Check if current character is not present in the dictionary
                pattern_dict[pattern[i]] = splitted_string[i]
            elif pattern_dict[pattern[i]] != splitted_string[i]:    # If character already present in dictionary, then check if it has same string value, if no then return False
                return False
            
            if splitted_string[i] not in s_dict:                    # Same logic as that of pattern
                s_dict[splitted_string[i]] = pattern[i]
            elif s_dict[splitted_string[i]] != pattern[i]:
                return False
            
        return True
        
# Approach: We follow the same approach as that of isomorphic strings. 
        